packages/backend-output-storage/src/stack_metadata_output_storage_strategy.ts

```diff
@@ -61,32 +78,46 @@ export class StackMetadataBackendOutputStorageStrategy
           `Metadata entry for ${keyName} at version ${existingMetadataEntry.version} already exists. Cannot add another entry for the same key at version ${version}.`
         );
       }
-    } else {
-      this.stack.addMetadata(keyName, {
-        version,
-        stackOutputs: Lazy.list({
-          produce: () => Array.from(listsMap ? listsMap.keys() : []),
-        }),
-      });
     }
 
-    Object.entries(backendOutputEntry.payload).forEach(([listName, value]) => {
-      if (!listsMap) {
-        listsMap = new Map();
-        this.lazyListValueMap.set(keyName, listsMap);
-      }
-      let outputList = listsMap.get(listName);
+    Object.entries(backendOutputEntry.payload ?? []).forEach(
+      ([listName, value]) => {
+        if (!value) {
+          return;
+        }
+        if (!listsMap) {
+          listsMap = new Map();
+          this.lazyListValueMap.set(keyName, listsMap);
+        }
+        let outputList = listsMap.get(listName);
 
-      if (outputList) {
-        outputList.push(value);
-      } else {
-        outputList = [value];
-        listsMap.set(listName, outputList);
+        if (outputList) {
+          outputList.push(value);
+        } else {
+          outputList = [value];
+          listsMap.set(listName, outputList);
 
-        new CfnOutput(this.stack, listName, {
-          value: Lazy.string({ produce: () => JSON.stringify(outputList) }),
-        });
+          new CfnOutput(this.stack, listName, {
+            value: Lazy.string({ produce: () => JSON.stringify(outputList) }),
+          });
+        }
       }
-    });
+    );
+
+    if (existingMetadataEntry) {
+      this.stack.addMetadata(keyName, {
+        version,
+        stackOutputs: Array.from(listsMap ? listsMap.keys() : []).concat(
+          existingMetadataEntry.stackOutputs.filter(
+            (output: string) => !listsMap?.has(output)
+          )
+        ),
+      });
```

Comment:
it seems this could also benefit from same simplification with `Set` .

And. This block could be lifted up to be closer to `existingMetadataEntry` definition above, so operations on metadata are toghether.

Also consider creating `private createOrAppendToMetadata()` method that is used in both `addBackendOutputEntry` and `append...` functions.

Reply:
Good catch. Done [9843705](https://github.com/aws-amplify/amplify-backend/pull/1742/commits/9843705aeb2cc5ba2894938175d691752ebae988)